deltacat/compute/compactor_v2/utils/primary_key_index.py

```diff
@@ -0,0 +1,196 @@
+from deltacat.compute.compactor.utils.primary_key_index import (
+    group_hash_bucket_indices as group_hash_bucket_indices_v1,
+)
+from deltacat.io.object_store import IObjectStore
+
+logger = logs.configure_deltacat_logger(logging.getLogger(__name__))
+
+
+def _append_sha1_hash_to_table(table: pa.Table, hash_column: pa.Array) -> pa.Table:
+    hash_column_np = hash_column.to_numpy()
+
+    result = []
+    for digest in hash_column_np:
+        result.append(hashlib.sha1(digest.encode("utf-8")).hexdigest())
+
+    return sc.append_pk_hash_string_column(table, result)
+
+
+def _is_sha1_desired(hash_column: pa.Array) -> bool:
+    return hash_column.nbytes > TOTAL_BYTES_IN_SHA1_HASH * len(hash_column)
+
+
+def group_by_pk_hash_bucket(
+    table: pa.Table, num_buckets: int, primary_keys: List[str]
+) -> np.ndarray:
+    table = generate_pk_hash_column(table, primary_keys, requires_sha1=True)
+
+    # group hash bucket record indices
+    result = group_record_indices_by_hash_bucket(
+        table,
+        num_buckets,
+    )
+
+    return result
+
+
+def generate_pk_hash_column(
+    table: pa.Table,
+    primary_keys: Optional[List[str]] = None,
+    requires_sha1: bool = False,
+) -> pa.Table:
+    """
+    Returns a new table after generating the primary key hash if desired.
+
+    1. If there are no primary keys, each hash will be unique uuid/sha1 hex
+    2. If there are more than 0 primary keys, returns a table with new columns appended.
+    """
+
+    start = time.monotonic()
+
+    can_sha1 = False
+    if primary_keys:
+        pk_columns = []
+        for pk_name in primary_keys:
+            pk_columns.append(pc.cast(table[pk_name], pa.string()))
+
+        pk_columns.append(PK_DELIMITER)
+        hash_column = pc.binary_join_element_wise(*pk_columns)
+
+        can_sha1 = requires_sha1 or _is_sha1_desired(hash_column)
+    else:
+        hash_column = pa.array(
+            [uuid.uuid4().hex for _ in range(len(table))], pa.string()
+        )
+
+    logger.info(
+        f"can_generate_sha1={can_sha1} for the table with hash column size"
+        f"={hash_column.nbytes} bytes, num_rows={len(hash_column)}, "
+        f"and requires_sha1={requires_sha1}"
+    )
+
+    if can_sha1:
+        table = _append_sha1_hash_to_table(table, hash_column)
+    else:
+        table = table.append_column(sc._PK_HASH_STRING_COLUMN_FIELD, hash_column)
+
+    end = time.monotonic()
+
+    logger.info(
+        f"Took {end - start}s to generate pk hash of len: {len(hash_column)}"
+        f" and size: {hash_column.nbytes} bytes"
+    )
+
+    return table
+
+
+def group_record_indices_by_hash_bucket(
+    pki_table: pa.Table, num_buckets: int
+) -> np.ndarray:
+    """
+    Groups the record indices by it's corresponding hash bucket. Hence, this method may
+    create num_buckets tables as a result.
+    """
+
+    input_table_len = len(pki_table)
+
+    hash_bucket_to_table = np.empty([num_buckets], dtype="object")
+    hash_bucket_id_col_list = []
```

Comment:
Minor: numpy array might be more suitable here

Reply:
Good point. This is not a minor comment:
```python
start = resource.getrusage(resource.RUSAGE_SELF).ru_maxrss
l = np.zeros(10000000)
p = pa.array(l)
end = resource.getrusage(resource.RUSAGE_SELF).ru_maxrss
end - start
# 35766272

l = []
for i in range(10000000):
    l.append(0)
p = pa.array(l)
end1 = resource.getrusage(resource.RUSAGE_SELF).ru_maxrss
end1 - end
169443328
```

So, there is a 4x memory savings that can be made just by using numpy arrays here.